### PR TITLE
Get WebKit MALLOC_HEAP_BREAKDOWN Configuration Building Again

### PR DESCRIPTION
**Pull request by stwrt**
#### 33b383131d5edd4552336cfb67114e4dddaaf363
<pre>
Get WebKit MALLOC_HEAP_BREAKDOWN Configuration Building Again
<a href="https://bugs.webkit.org/show_bug.cgi?id=259786">https://bugs.webkit.org/show_bug.cgi?id=259786</a>

Reviewed by Elliott Williams.

Get the Malloc Heap Breakdown build compiling again. Lots of header includes and exporting of symbols.
This really should not touch the system in any other way, but it does modify files across numerous components.

* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
* Source/JavaScriptCore/assembler/LinkBuffer.h:
* Source/JavaScriptCore/bytecode/InstructionStream.cpp:
* Source/JavaScriptCore/runtime/StringReplaceCache.h:
* Source/JavaScriptCore/runtime/StringSplitCache.h:
* Source/WTF/wtf/BitVector.cpp:
* Source/WTF/wtf/DebugHeap.h:
* Source/WTF/wtf/FastMalloc.h:
* Source/WTF/wtf/MetaAllocator.cpp:
* Source/WTF/wtf/text/CString.h:
* Source/WTF/wtf/text/cf/StringImplCF.cpp:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBindGroupLayout.h:
* Source/WebGPU/WGSL/AST/ASTBuilder.h:
* Source/WebKit/Shared/ShareableBitmap.cpp:
* Source/WebKit/Shared/ShareableBitmap.h:
* Source/WebKit/Shared/ShareableBitmapHandle.cpp:
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:

Canonical link: <a href="https://commits.webkit.org/266981@main">https://commits.webkit.org/266981@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/d92b7914c96bb7500b12259258804912cb26435e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15290 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15955 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14352 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15430 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18110 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15693 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15473 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15903 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13002 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17779 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13182 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13795 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/20747 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/13104 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14263 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13962 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17222 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14537 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14531 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/12311 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/15483 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13801 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3940 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3671 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18148 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/15718 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14364 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3752 "Passed tests") |

### CODE REVIEW

File: Source/JavaScriptCore/assembler/LinkBuffer.h [hunk 1]
```diff
@@ -39,10 +39,13 @@
 #include "MacroAssemblerCodeRef.h"
 #include <wtf/DataLog.h>
 #include <wtf/FastMalloc.h>
+#include <wtf/NeverDestroyed.h>
 #include <wtf/Noncopyable.h>
 
 namespace JSC {
 
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(BranchCompactionLinkBuffer);
```
Comment (rreno) [inline Source/JavaScriptCore/assembler/LinkBuffer.h]: This should stay in the .cpp since the class is wholly contained there.

File: Source/JavaScriptCore/runtime/StringReplaceCache.h [hunk 2]
```diff
@@ -30,14 +30,11 @@
...
 
-DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(StringReplaceCache);
-
 class JSImmutableButterfly;
 class RegExp;
 
 class StringReplaceCache {
-    WTF_MAKE_FAST_ALLOCATED;
-    WTF_MAKE_NONCOPYABLE(StringReplaceCache);
+
```
Comment (rreno) [inline Source/JavaScriptCore/runtime/StringReplaceCache.h:41]: This seems unnecessary.

File: Source/JavaScriptCore/runtime/StringSplitCache.h [hunk 3]
```diff
@@ -30,13 +30,9 @@
 
 namespace JSC {
 
-DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(StringSplitCache);
-
 class JSImmutableButterfly;
 
 class StringSplitCache {
-    WTF_MAKE_FAST_ALLOCATED;
```
Comment (rreno) [inline Source/JavaScriptCore/runtime/StringSplitCache.h]: Ditto StringReplaceCache. Why are you removing the noncopyable/fast malloc macros?

File: Source/WTF/wtf/BitVector.h [hunk 4]
```diff
@@ -45,6 +45,8 @@ namespace WTF {
 
 class FixedBitVector;
 
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(BitVector);
```
Comment (rreno) [inline Source/WTF/wtf/BitVector.h]: You didn't remove the WTF_MAKE_FAST_ALLOCATED from the class definition.

File: Source/WTF/wtf/MetaAllocator.h [hunk 5]
```diff
@@ -41,6 +41,8 @@ namespace WTF {
 
 #define ENABLE_META_ALLOCATOR_PROFILE 0
 
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(MetaAllocatorFreeSpace);
```
Comment (rreno) [inline Source/WTF/wtf/MetaAllocator.h]: This should probably have the same name as the nested class (i.e. `MetaAllocatorFreeSpaceNode`).

Comment (rreno) [reply Source/WTF/wtf/MetaAllocator.h]: I understand that it didn't before your patch.

File: Source/WTF/wtf/text/StringImpl.h [hunk 6]
```diff
@@ -63,6 +63,9 @@ class LLIntOffsetsExtractor;
 
 namespace WTF {
 
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(StringWrapperCFAllocator);
```
Comment (rreno) [inline Source/WTF/wtf/text/StringImpl.h]: Why did this need to move to the header?

File: Source/WTF/wtf/text/cf/StringImplCF.cpp [hunk 7]
```diff
@@ -32,10 +32,9 @@
 
 namespace WTF {
 
-namespace StringWrapperCFAllocator {
+DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(StringWrapperCFAllocator);
```
Comment (rreno) [inline Source/WTF/wtf/text/cf/StringImplCF.cpp]: Why did this move out of the namespace?

File: Source/WebGPU/WGSL/AST/ASTBuilder.h [hunk 8]
```diff
@@ -46,10 +48,10 @@ class ShaderModule;
 
 namespace AST {
 
-class Node;
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(WGSLAST, WTF_INTERNAL);
 
-DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(WGSLAST);
 
+class Node;
```
Comment (rreno) [inline Source/WebGPU/WGSL/AST/ASTBuilder.h]: Unnecessary change.

File: Source/WebKit/Shared/ShareableBitmap.h [hunk 9]
```diff
@@ -116,7 +119,10 @@ class ShareableBitmapHandle  {
     ShareableBitmapConfiguration m_configuration;
 };
 
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(ShareableBitmap, WTF_INTERNAL);
+
 class ShareableBitmap : public ThreadSafeRefCounted<ShareableBitmap> {
+    WTF_MAKE_FAST_ALLOCATED;
```
Comment (rreno) [inline Source/WebKit/Shared/ShareableBitmap.h]: Why this and not _WITH_HEAP_IDENTIFIER?

File: Source/WebKit/Shared/ShareableBitmap.h [hunk 10]
```diff
@@ -93,6 +95,7 @@ class ShareableBitmapConfiguration {
 };
 
 class ShareableBitmapHandle  {
+    WTF_MAKE_FAST_ALLOCATED;
```
Comment (rreno) [inline Source/WebKit/Shared/ShareableBitmap.h]: Why?

**Review by rreno — CHANGES_REQUESTED** [review]

File: Source/JavaScriptCore/runtime/StringReplaceCache.h (hunk 2 above)
Comment (stwrt) [reply Source/JavaScriptCore/runtime/StringReplaceCache.h:41]: String Replace Cache was having some issues with compiling; need to look into the again. I noticed some of the other similar caches did not use the FAST_ALLOCATED, so I removed this.

File: Source/JavaScriptCore/assembler/LinkBuffer.h (hunk 1 above)
Comment (stwrt) [reply Source/JavaScriptCore/assembler/LinkBuffer.h]: Get this error if I keep I keep it in the cpp file.

```
JavaScriptCore GenerateTAPI JavaScriptCore.tbd
warning: violations found for arm64e
error: no declaration found for exported symbol 'JSC::BranchCompactionLinkBufferMalloc::debugHeap()' in dynamic library
WebGPU CompileC Adapter.o
WebGPU CompileC CommandEncoder.o
```

Comment (emw-apple) [reply Source/JavaScriptCore/assembler/LinkBuffer.h]: If it stays in the class, it needs to have visibility hidden, because otherwise you're exporting the debugHeap() symbol without putting a declaration in headers for a client to import.

Comment (rreno) [reply Source/JavaScriptCore/assembler/LinkBuffer.h]: Yep so you should pass `WTF_PRIVATE` as the second parameter to the DECLARE macro.

Comment (rreno) [reply Source/JavaScriptCore/assembler/LinkBuffer.h]: jinx

File: Source/WTF/wtf/DebugHeap.h [hunk 11]
```diff
@@ -55,9 +55,11 @@ class DebugHeap {
 #endif
 };
 
-#define DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(Type) \
+#define DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(Type) DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(Type, WTF_EXPORT_PRIVATE)
+
+#define DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(Type, Export) \
```
Comment (rreno) [inline Source/WTF/wtf/DebugHeap.h:60]: You'll also need this macro available in non MHB builds.

File: Source/JavaScriptCore/runtime/StringSplitCache.h (hunk 3 above)
Comment (justinmichaud) [reply Source/JavaScriptCore/runtime/StringSplitCache.h]: Worth noting: this will almost certainly have perf implications

File: Source/JavaScriptCore/assembler/LinkBuffer.h (hunk 1 above)
Comment (stwrt) [reply Source/JavaScriptCore/assembler/LinkBuffer.h]: Changed to `DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(BranchCompactionLinkBuffer, WTF_INTERNAL);
` and it builds successfully now.

File: Source/JavaScriptCore/runtime/StringReplaceCache.h (hunk 2 above)
Comment (stwrt) [reply Source/JavaScriptCore/runtime/StringReplaceCache.h:41]: Re-added. Was able to resolve with `DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT`.

File: Source/JavaScriptCore/runtime/StringSplitCache.h (hunk 3 above)
Comment (stwrt) [reply Source/JavaScriptCore/runtime/StringSplitCache.h]: Added back.

File: Source/WTF/wtf/BitVector.h (hunk 4 above)
Comment (stwrt) [reply Source/WTF/wtf/BitVector.h]: Where is WTF_MAKE_FAST_ALLOCATED in the .cpp file?

File: Source/WTF/wtf/DebugHeap.h (hunk 11 above)
Comment (stwrt) [reply Source/WTF/wtf/DebugHeap.h:60]: Right. Need to add support for the non MHB #define

File: Source/WTF/wtf/MetaAllocator.h (hunk 5 above)
Comment (stwrt) [reply Source/WTF/wtf/MetaAllocator.h]: This can be handled in a future patch then. It falls outside the current patch's scope.

File: Source/JavaScriptCore/assembler/LinkBuffer.h [hunk 12]
```diff
@@ -39,6 +39,7 @@
 #include "MacroAssemblerCodeRef.h"
 #include <wtf/DataLog.h>
 #include <wtf/FastMalloc.h>
+#include <wtf/NeverDestroyed.h>
```
Comment (emw-apple) [inline Source/JavaScriptCore/assembler/LinkBuffer.h]: Seems like this include should be in `DebugHeap.h`, since that's where the macro is defined.

File: Source/JavaScriptCore/bytecode/InstructionStream.cpp [hunk 13]
```diff
@@ -26,6 +26,8 @@
 #include "config.h"
 #include "InstructionStream.h"
 
+#include <wtf/NeverDestroyed.h>
```
Comment (emw-apple) [inline Source/JavaScriptCore/bytecode/InstructionStream.cpp:29]: Ditto.

File: Source/JavaScriptCore/runtime/StringSplitCache.h [hunk 14]
```diff
@@ -27,10 +27,11 @@
 #pragma once
 
 #include <array>
+#include <wtf/FastMalloc.h>
```
Comment (emw-apple) [inline Source/JavaScriptCore/runtime/StringSplitCache.h]: Ditto.

File: Source/WTF/wtf/DebugHeap.h [hunk 15]
```diff
@@ -88,7 +90,9 @@ class DebugHeap {
 
 #else // ENABLE(MALLOC_HEAP_BREAKDOWN)
 
-#define DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(Type) \
+#define DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(Type) DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(Type, nullptr)
```
Comment (emw-apple) [inline Source/WTF/wtf/DebugHeap.h]: `nullptr` is weird, since this parameter is never used like a pointer. I think I would just put WTF_EXPORT_PRIVATE here? (That would also enable you to hoist this definition outside of the `#ifdef` and only define it once, if you'd like.)

File: Source/WTF/wtf/text/CString.h [hunk 16]
```diff
@@ -26,6 +26,7 @@
 #pragma once
 
 #include <span>
+#include <wtf/FastMalloc.h>
```
Comment (emw-apple) [inline Source/WTF/wtf/text/CString.h]: Ditto.

File: Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBindGroupLayout.h [hunk 17]
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include <wtf/FastMalloc.h>
```
Comment (emw-apple) [inline Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBindGroupLayout.h]: Ditto.

File: Source/WebGPU/WGSL/AST/ASTBuilder.h [hunk 18]
```diff
@@ -25,7 +25,9 @@
 
 #pragma once
 
+#include <wtf/FastMalloc.h>
 #include <wtf/MallocPtr.h>
+#include <wtf/NeverDestroyed.h>
```
Comment (emw-apple) [inline Source/WebGPU/WGSL/AST/ASTBuilder.h]: Ditto.

File: Source/WebKit/Shared/ShareableBitmap.h [hunk 19]
```diff
@@ -31,6 +31,8 @@
 #include <WebCore/IntRect.h>
 #include <WebCore/PlatformImage.h>
 #include <wtf/ArgumentCoder.h>
+#include <wtf/ExportMacros.h>
+#include <wtf/FastMalloc.h>
```
Comment (emw-apple) [inline Source/WebKit/Shared/ShareableBitmap.h]: Ditto.

**Review by emw-apple — COMMENTED** [review]
The proliferation of `WTF_INTERNAL` outside of WTF is unfortunate. But I don't think we have a good alternative, unless we were to refactor `DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER` to **not** default to public visibility. Please feel free to do that, but I'm not going to block the PR on it.

Otherwise, looks good to me except for a few nits:

File: Source/WebGPU/WGSL/AST/ASTBuilder.h (hunk 18 above)
Comment (stwrt) [reply Source/WebGPU/WGSL/AST/ASTBuilder.h]:
```
WGSL CompileC ASTBuilder.o
/Users/bts/dev/OpenSource/Source/WebGPU/WGSL/AST/ASTBuilder.cpp:33:1: error: implicit instantiation of undefined template 'WTF::LazyNeverDestroyed<WTF::DebugHeap>'
DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(WGSLAST);
^
In file included from /Users/bts/dev/OpenSource/Source/WebGPU/WGSL/AST/ASTBuilder.cpp:27:
In file included from /Users/bts/dev/OpenSource/Source/WebGPU/WGSL/AST/ASTBuilder.h:28:
/Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/DebugHeap.h:82:51: note: expanded from macro 'DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER'
        static LazyNeverDestroyed<WTF::DebugHeap> heap; \
In file included from /Users/bts/dev/OpenSource/Source/WebGPU/WGSL/AST/ASTBuilder.cpp:27:
In file included from /Users/bts/dev/OpenSource/Source/WebGPU/WGSL/AST/ASTBuilder.h:32:
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/Vector.h:33:
/Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/Forward.h:78:61: note: template is declared here
template<typename, typename = AnyThreadsAccessTraits> class LazyNeverDestroyed;
1 error generated.
```

File: Source/JavaScriptCore/bytecode/InstructionStream.cpp (hunk 13 above)
Comment (stwrt) [reply Source/JavaScriptCore/bytecode/InstructionStream.cpp:29]:
```
JavaScriptCore CompileC UnifiedSource33.o
JavaScriptCore CompileC UnifiedSource32.o
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/DerivedSources/JavaScriptCore/unified-sources/UnifiedSource32.cpp:1:
./bytecode/InstructionStream.cpp:33:1: error: implicit instantiation of undefined template 'WTF::LazyNeverDestroyed<WTF::DebugHeap>'
DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(InstructionStream);
^
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/DerivedSources/JavaScriptCore/unified-sources/UnifiedSource32.cpp:1:
In file included from ./bytecode/InstructionStream.cpp:26:
In file included from /Users/bts/dev/OpenSource/Source/JavaScriptCore/config.h:39:
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/FastMalloc.h:25:
/Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/DebugHeap.h:82:51: note: expanded from macro 'DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER'
        static LazyNeverDestroyed<WTF::DebugHeap> heap; \
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/DerivedSources/JavaScriptCore/unified-sources/UnifiedSource32.cpp:1:
In file included from ./bytecode/InstructionStream.cpp:27:
In file included from ./bytecode/InstructionStream.h:29:
In file included from ./bytecode/BytecodeIndex.h:28:
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/HashTraits.h:25:
/Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/Forward.h:78:61: note: template is declared here
template<typename, typename = AnyThreadsAccessTraits> class LazyNeverDestroyed;
1 error generated.
```

**Review by emw-apple — APPROVED** [review]
